### DISCUSSION TITLE
Running PrairieLearn in Docker with unprivileged user

### DISCUSSION THREAD

**Discussion by jonatanschroeder** [PrairieLearn/PrairieLearn#10661] [Dev]
One issue I've had for a long time is that, when running PrairieLearn inside docker, the PrairieLearn process runs as root. This causes any files and directories created by PrairieLearn (or related tools running inside the container) to be owned by root, which can cause issues if trying to access these files outside the container.

I managed to find a workaround, which is to create a local user inside the container and run everything (except Postgres, which requires its own user) as the user. The command I use is (with some simplifications):

```bash
docker run -it --name pl --rm -p 3000:3000 -v $HOME/pl:/PrairieLearn -v $HOME/mycoursedir:/course \
  -v /tmp/pl_ag_jobs:/jobs -e HOST_JOBS_DIR=/tmp/pl_ag_jobs -v /var/run/docker.sock:/var/run/docker.sock \
  --add-host=host.docker.internal:172.17.0.1 -w /PrairieLearn -v $HOME/pldb:/var/postgres \
  prairielearn/prairielearn bash -c "adduser --uid $(id -u) pl ; groupadd --gid $(getent group docker | cut -d: -f3) docker ; usermod -aG docker pl ; make start-postgres ; mkdir -p s3rver ; chown pl:pl s3rver ; runuser -u pl -- make dev-all"
```

This assumes that I'm running PrairieLearn from source, if that is not the case you can remove `-v $HOME/pl:/PrairieLearn`. Also, replace `$HOME/mycoursedir` with the directory of your course (and add course2/3/etc. as needed).

This works for most PrairieLearn functionality. By itself, though, it doesn't properly work with workspaces. In order to get that to work, I need to set, in config.json (both in apps/prairielearn and apps/workspace-host):

```json
{
  "workspaceJobsDirectoryOwnerUid": 1000,
  "workspaceJobsDirectoryOwnerGid": 1000
}
```

Doing this change also works as a workaround for #9828.

I hope this helps others as well. If @nwalters512 or @mwest1066 would prefer to have a PL to have this handled by default, I'm open to creating a PR, but I believe it's a bit of a niche use case.

**Comment by jonatanschroeder** [comment]
> If @nwalters512 or @mwest1066 would prefer to have a PL to have this handled by default, I'm open to creating a PR

To clarify: the PR could change the image to allow it to more easily run with [the `--user` option of `docker run`](https://docs.docker.com/engine/containers/run/#user). In essence, this would mean:
* Create a generous sudoers file;
* On docker/init.sh, if the current user is not root, run using sudo:
  * create a user with the current UID/GID;
  * create docker group (if it doesn't exist) based on group owner of /var/run/docker.sock, and add new user to the group;
* Change postgres start script to run with sudo;
* Change s3rver's default directory to a location that doesn't cause permission problems, or something similar;
* Change the default config options for workspace directory owner to the current user, if not provided;
  * Alternative: change the chown call in workspaces to not run if the config option is not provided.

I don't know if any of the options above has a security concern, but considering we're running as root at the moment I don't think this should be a problem. This functionality could be somewhat simplified if we can assume that the user has a UID of 1000 and that docker is group 1001, but I don't think that's safe to assume.

I'm open to alternatives if you want to consider. And again reiterating, I don't mind just keeping the workaround above in my environment, if you believe this is not worth adding to PL itself.

**Reply by joelostblom** [reply]
This sounds like a convenience improvement to me! I initially tried appending `--user $(id -u):$(id -g)` to the docker command in the PL documentation but noticed that was not sufficient and would not have been able to figure it out without jonatan's post here and help on slack

**Reply by echuber2** [reply]
Regarding workspaces, I've been trying to add entrypoints to workspaces (currently in the VS Code and RStudio ones, PR not merged for Jupyter yet) where it uses gosu to change from 0 to 1001 if the user is running locally with the default of starting everything as root. If a new feature would let you run PL itself as a different user, maybe the whole issue of how workspaces run locally could be revisited. See also: https://github.com/PrairieLearn/PrairieLearn/discussions/8897

**Reply by joelostblom** [reply]
@jonatanschroeder Did it work for you to run workspaces locally with the config.json modification? I just tried it and I while JupyterLab launches, I run into permission errors:

![Screenshot from 2024-09-24 09-24-08](https://github.com/user-attachments/assets/55dc6a3d-d6f6-4a5f-9c19-0d65c11d192c)
![Screenshot from 2024-09-24 09-24-01](https://github.com/user-attachments/assets/fba34ee1-9c69-4e86-9f01-82e7957cbc01)



With RStudio, I see this instead:

![image](https://github.com/user-attachments/assets/6ccdd2d1-2708-450c-9d9f-a89fd28b5f8b)

Trying to change `1000` to `1001` as per the config.json from @echuber2 in https://github.com/PrairieLearn/PrairieLearn/discussions/8897 gives this for JupyterLab:

![image](https://github.com/user-attachments/assets/4f10b056-4795-42be-baae-304d4df0119f)

And this for RStudio:

![image](https://github.com/user-attachments/assets/af86cef0-4819-47ba-9eca-0dbe965e8205)

I also tried with the config set to `0` (and empty) but then I get permission errors. The docker command I run is:

```bash
docker run -it --rm -p 3000:3000 --name pl \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -v ${HOME}/PrairieLearn/pl_ag_jobs:/jobs \
    -e HOST_JOBS_DIR=/tmp/pl_ag_jobs \
    --add-host=host.docker.internal:172.17.0.1 \
    -v ${PWD}:/course \
    -v ${HOME}/PrairieLearn/pldb:/var/postgres \
    -v ${HOME}/PrairieLearn/config.json:/PrairieLearn/config.json \
    -w /PrairieLearn \
    prairielearn/prairielearn \
    bash -c "adduser --uid $(id -u) pl ; groupadd --gid $(getent group docker | cut -d: -f3) docker ; usermod -aG docker pl ; make start-postgres ; mkdir s3rver; chown pl:pl s3rver; runuser -u pl -- make dev-all"
```

and when I run it for the first time with an empty database in `pldb`, I do see the following error message on startup (but it is not present after relaunching the container):

```
error: Error initializing workspace host: relation "workspace_hosts" does not exist {"code":"42P01","data":{"sql":"-- BLOCK insert_workspace_hosts\nINSERT INTO\n  workspace_hosts (\n    instance_id,\n    hostname,\n    state,\n    state_changed_at,\n    ready_at\n  )\nVALUES\n  ($1, $2, 'ready', NOW(), NOW())\nON CONFLICT (instance_id) DO\nUPDATE\nSET\n  hostname = EXCLUDED.hostname,\n  state = EXCLUDED.state,\n  state_changed_at = EXCLUDED.state_changed_at,\n  ready_at = EXCLUDED.ready_at;\n","sqlError":{"code":"42P01","file":"parse_relation.c","length":114,"line":"1392","message":"relation \"workspace_hosts\" does not exist","name":"error","position":"47","routine":"parserOpenTable","severity":"ERROR"},"sqlParams":{"hostname":"localhost:8081","instance_id":"devWSHost1"}},"file":"parse_relation.c","length":114,"line":"1392","name":"error","position":"47","routine":"parserOpenTable","severity":"ERROR","stack":"error: relation \"workspace_hosts\" does not exist\n    at /PrairieLearn/node_modules/pg/lib/client.js:526:17\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async PostgresPool.queryWithClientAsync (/PrairieLearn/packages/postgres/src/pool.ts:319:22)\n    at async PostgresPool.queryAsync (/PrairieLearn/packages/postgres/src/pool.ts:534:14)\n    at async async.series.sqldb.queryAsync.instance_id (file:///PrairieLearn/apps/workspace-host/src/interface.js:204:7)"}
error: Could not mark self as unhealthy relation "workspace_hosts" does not exist {"code":"42P01","data":{"sql":"-- BLOCK mark_host_unhealthy\nWITH\n  updated_workspace_hosts AS (\n    UPDATE workspace_hosts AS wh\n    SET\n      state = 'unhealthy',\n      state_changed_at = NOW(),\n      unhealthy_at = NOW(),\n      unhealthy_reason = $1\n    WHERE\n      wh.instance_id = $2\n      AND wh.state IN ('launching', 'ready', 'draining')\n    RETURNING\n      wh.id,\n      wh.state,\n      wh.unhealthy_reason\n  )\nINSERT INTO\n  workspace_host_logs (workspace_host_id, state, message)\nSELECT\n  wh.id,\n  wh.state,\n  wh.unhealthy_reason\nFROM\n  updated_workspace_hosts AS wh;\n","sqlError":{"code":"42P01","file":"parse_relation.c","length":114,"line":"1392","message":"relation \"workspace_hosts\" does not exist","name":"error","position":"77","routine":"parserOpenTable","severity":"ERROR"},"sqlParams":{"instance_id":"devWSHost1","unhealthy_reason":{"code":"42P01","data":{"sql":"-- BLOCK insert_workspace_hosts\nINSERT INTO\n  workspace_hosts (\n    instance_id,\n    hostname,\n    state,\n    state_changed_at,\n    ready_at\n  )\nVALUES\n  ($1, $2, 'ready', NOW(), NOW())\nON CONFLICT (instance_id) DO\nUPDATE\nSET\n  hostname = EXCLUDED.hostname,\n  state = EXCLUDED.state,\n  state_changed_at = EXCLUDED.state_changed_at,\n  ready_at = EXCLUDED.ready_at;\n","sqlError":{"code":"42P01","file":"parse_relation.c","length":114,"line":"1392","message":"relation \"workspace_hosts\" does not exist","name":"error","position":"47","routine":"parserOpenTable","severity":"ERROR"},"sqlParams":{"hostname":"localhost:8081","instance_id":"devWSHost1"}},"file":"parse_relation.c","length":114,"line":"1392","name":"error","position":"47","routine":"parserOpenTable","severity":"ERROR"}}},"file":"parse_relation.c","length":114,"line":"1392","name":"error","position":"77","routine":"parserOpenTable","severity":"ERROR","stack":"error: relation \"workspace_hosts\" does not exist\n    at /PrairieLearn/node_modules/pg/lib/client.js:526:17\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at async PostgresPool.queryWithClientAsync (/PrairieLearn/packages/postgres/src/pool.ts:319:22)\n    at async PostgresPool.queryAsync (/PrairieLearn/packages/postgres/src/pool.ts:534:14)\n    at async markSelfUnhealthy (file:///PrairieLearn/apps/workspace-host/src/interface.js:408:5)\n    at async file:///PrairieLearn/apps/workspace-host/src/interface.js:263:5"}
```

**Reply by jonatanschroeder** [reply]
I haven't run into issues locally with Jupyterlab (though I don't use it often). I don't have experience with RStudio.

**Reply by echuber2** [reply]
If you're seeing issues in both Jupyter and Rstudio, maybe you can ping me/us on Slack and we can talk through it a bit more in detail. In the container error log you showed that mentions user 1000, I put that in defensively as it's meant to run as either 0 (which steps down to 1001 automatically) or be run as 1001 explicitly. If you're trying to run it as 1000 then it won't work as intended. It's possible that some of the tips Jonatan gave simply conflict with the tips I gave elsewhere.

**Reply by joelostblom** [reply]
Following up with the same comment I left on slack in case someone stumbles over this. I believe I have solved my issue by adding chown -R pl:pl /jobs to the docker command from Jonatan. I updated it here as well https://github.com/PrairieLearn/PrairieLearn/discussions/10661#discussioncomment-10648973. It works well for jupyterlab, but I think I'm running into an unrelated issue with Rstudio, so I will stick with running R and Py in jlab.

**Comment by joelostblom** [comment]
Thank you for posting this @jonatanschroeder ! As someone who is new to PL, I found it difficult to work locally with the docker image due to files created in the UI being owned by root, and having to manually `chown` to change that. This tweak to the command makes it more convenient to work locally with PL for me. 

For others who are new to PL and come across this thread, here is the command with an (likely imperfect) explanation for what every line does:

```bash
# For details, see https://github.com/PrairieLearn/PrairieLearn/discussions/10661
docker run -it --rm -p 3000:3000 --name pl \
    `# These optional four lines enable external workspaces and graders (also needs a file edit, see link above)` \
    `# For details, see https://prairielearn.readthedocs.io/en/latest/installing/#support-for-external-graders-and-workspaces` \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -v ${HOME}/PrairieLearn/pl_ag_jobs:/jobs \
    -e HOST_JOBS_DIR=${HOME}/PrairieLearn/pl_ag_jobs \
    --add-host=host.docker.internal:172.17.0.1 \
    `# Optional to copy over a local config file, e.g. for workspaces as in https://github.com/PrairieLearn/PrairieLearn/discussions/10661#discussion-7183259` \
    -v ${HOME}/PrairieLearn/config.json:/PrairieLearn/config.json \
    `# Optional to keep test data such as dummy users between docker restarts` \
    -v ${HOME}/PrairieLearn/pldb:/var/postgres \
    `# Make the current working dir available as a course in "Load from disk"` \
    -v ${PWD}:/course \
     `# Set the docker working directory to where the MAKEFILE is` \
    -w /PrairieLearn \
    `# Specify which image to run` \
    prairielearn/prairielearn \
    `# Replicate your local user in the docker image so that files created are own by you instead of root` \
    bash -c "adduser --uid $(id -u) pl ; groupadd --gid $(getent group docker | cut -d: -f3) docker ; usermod -aG docker pl ; make start-postgres ; mkdir s3rver ; chown pl:pl s3rver ; chown -R pl:pl /jobs ; runuser -u pl -- make dev-all"

```

**Comment by mwest1066** [comment]
> I hope this helps others as well. If @nwalters512 or @mwest1066 would prefer to have a PL to have this handled by default, I'm open to creating a PR, but I believe it's a bit of a niche use case.

Part of the problem here is that both @nwalters512 and I always run PL natively, and we never use docker, so we aren't personally running into any of the issues that make this kind of setup helpful.

I'll defer to @nwalters512 on adding this support into the main PL docker build itself.

**Comment by trombonekenny** [comment]
I would be willing to collaborate on a PR that made PrairieLearn dev docker run well with `--user`. That's my build environment and it would be nice not to have `root` files all over my dev paths.
